### PR TITLE
Add Typescript

### PR DESCRIPTION
**Pull request by marvin-j97**
**WIP!**

#### Done:

1. Add typescript & D3 types to dev dependencies
2. Add tsconfig
3. Change project structure (dist now contains compiled JS, src will contain TS code)
4. Add typings filed to package.json
5. Rewritten vector.js to vector.ts

#### Todo:

1. Convert JS to TS
2. Adjust tsconfig to taste

### Things to consider:

- The build needs to be compiled before release (run `npx tsc`)
- Currently the build is not bundled and written as ES modules, it may be good to add a separate ES5/CommonJS bundle
- ES6 private fields only really work with ES6, otherwise they are compiled into weak maps for backwards compat by the TS compiler - IMO, using TS, ES6 private fields are not needed and using the *private* keyword is enough
-  There are some weird lines in the code like `d3.select()` that are, at least according to D3's types, illegal invocations (an argument is always required)
- Add linting and auto formatting rules (prettier)
- Add automatic testing pipeline (verify TS build, linting, additional tests)
- Documentation may be autogenerated by something like [typedoc](https://typedoc.org/)

### CODE REVIEW

**Comment by imaphatduc** [comment]
I've seen your PR and would like to thank you for your work :). I'm currently working on migrating to TS too, so I think you should push your codes into `src-ts/` directory in the repo's `src-typescript` branch.

**Review by imaphatduc — APPROVED** [review]

**Comment by imaphatduc** [comment]
Why did you use `ElementMapFunction` type in `vector.ts` file instead of the normal `Function`? I guess this will force the type to accept only one parameter, or maybe I've missed something.

**Comment by marvin-j97** [comment]
> Why did you use `ElementMapFunction` type in `vector.ts` file instead of the normal `Function`? I guess this will force the type to accept only one parameter, or maybe I've missed something.

Yes, it's a strict function type (https://www.typescriptlang.org/docs/handbook/2/functions.html).
Actually, typescript-eslint would warn you when using 'Function' because:

```
Avoid the Function type, as it provides little safety for the following reasons:
- It provides no type safety when calling the value, which means it's easy to provide the wrong arguments.
- It accepts class declarations, which will fail when called, as they are called without the new keyword.
```

**Comment by marvin-j97** [comment]
I've merged your typescript branch
I recommend using

```
noImplicitAny
strictNullChecks
strictFunctionTypes
strictBindCallApply
strictPropertyInitialization
noImplicitThis
alwaysStrict
noUnusedLocals
noUnusedParameters
exactOptionalPropertyTypes
noFallthroughCasesInSwitch
```

in the tsconfig, once the rewrite is done.

**Comment by imaphatduc** [comment]
> 
> Yes, it's a strict function type (https://www.typescriptlang.org/docs/handbook/2/functions.html). Actually, typescript-eslint would warn you when using 'Function' because:
> 
> ```
> Avoid the Function type, as it provides little safety for the following reasons:
> - It provides no type safety when calling the value, which means it's easy to provide the wrong arguments.
> - It accepts class declarations, which will fail when called, as they are called without the new keyword.
> ```
OK, I got it!

File: src-ts/cubicons/coordinateSys.ts [hunk 1]
```diff
@@ -1,10 +1,11 @@
 import * as d3 from "d3";
```
Comment (imaphatduc) [inline src-ts/cubicons/coordinateSys.ts:1]: I think we should get rid of `import * as d3 from "d3"` lines and import each module specifically.

File: package.json [hunk 2]
```diff
@@ -14,11 +14,16 @@
         "katex": "^0.15.1"
     },
     "devDependencies": {
+        "@types/d3": "^7.1.0",
+        "@types/d3-scale": "^4.0.2",
```
Comment (imaphatduc) [inline package.json:18]: Why did you include both `d3` and `d3-scale` in this file?

File: package.json [hunk 3]
```diff
@@ -14,11 +14,16 @@
...
+        "@types/katex": "^0.11.1",
+        "typescript": "^4.5.4",
         "vite": "^2.7.4"
     },
     "description": "Play eye-catching math visualization on the web using SVG",
-    "module": "src/index.js",
-    "main": "src/index.js",
+    "module": "dist/index.js",
+    "main": "dist/index.js",
```
Comment (imaphatduc) [inline package.json:25]: I included `dist/` dir in `.gitignore` file, so should we set entry point to `src/` dir? I don't clearly understand about TypeScript entry point, but I can guess that that file should be `index.d.ts` in the `src/` dir, shouldn't it?

File: package.json [hunk 4]
```diff
@@ -14,11 +14,16 @@
...
+    "types": "dist/index.d.ts",
```
Comment (imaphatduc) [inline package.json:26]: `dist/` dir is in `.gitignore` file.

**Review by imaphatduc — COMMENTED** [review]
- Your implementations in `vector.ts` and `cubecubed.ts` look good, I'll test things out a little bit and merge them as soon as you confirm that this PR is ready.
- However, in `package.json` file, I can't see why you include both d3 and d3-scale as dev dependencies.
- In `.ts` files, I considered your ideas about import each module at a time (not `import * as d3 from "d3"`). I've included all the necessary import lines and you don't need to do it any more.
**Important**:
I removed all the `.js` files in `src/` directory (`src-typescript` branch) and moved `.ts` files in there, so you should pull my `src-typescript` branch to avoid conflicts.
Thank you.

**Comment by marvin-j97** [comment]
I've decided to split the PR
https://github.com/imaphatduc/cubecubed/pull/4
https://github.com/imaphatduc/cubecubed/pull/5

**Comment by imaphatduc** [comment]
**Just to leave a note**: All the TS files that I've converted from JS work fine (at least in my computer). I think everything left to do in this PR should be resolving the entry point `index.d.ts`, and then I'll merge it to the `master` branch. In my opinion, you should:
- remove `drawGrid.ts`, `coordinateSys.ts` and `text.ts` (`package.json` and `package-lock.json` too) from the PR (as I added all the necessary modules).
- resolve the entry point.
- include your modified `vector.ts` and `cubecubed.ts` in another PR (my suggestion for the name is "Implement types in `vector.ts` and `cubecubed.ts` file")
---
If you think some TS files need improvements, consider making other PRs and I'll check through them all.

**Pull request closed without merging**